common/network-types/src/session/protocol.rs

```diff
@@ -0,0 +1,403 @@
+    fn from(value: SessionMessage<C>) -> Self {
+        let mut ret = Vec::with_capacity(C);
+        ret.push(SessionMessage::<C>::VERSION);
+        ret.push(SessionMessageDiscriminants::from(&value) as u8);
+
+        match value {
+            SessionMessage::Segment(s) => ret.extend(Vec::from(s)),
+            SessionMessage::Request(b) => ret.extend(Vec::from(b)),
+            SessionMessage::Acknowledge(f) => ret.extend(Vec::from(f)),
+        };
+
+        ret.shrink_to_fit();
+        ret
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::session::Frame;
+    use bitvec::bitarr;
+    use hex_literal::hex;
+    use hopr_platform::time::native::current_time;
+    use rand::{thread_rng, Rng};
+    use std::time::SystemTime;
+
+    #[test]
+    fn test_session_message_segment() -> anyhow::Result<()> {
+        const SEG_SIZE: usize = 8;
+
+        let mut segments = Frame {
+            frame_id: 10,
+            data: hex!("deadbeefcafebabe").into(),
+        }
+        .segment(SEG_SIZE)?;
+
+        const MTU: usize = SEG_SIZE + Segment::HEADER_SIZE + 2;
+
+        let msg_1 = SessionMessage::<MTU>::Segment(segments.pop().unwrap());
+        let data = Vec::from(msg_1.clone());
+        let msg_2 = SessionMessage::try_from(&data[..]).unwrap();
+
+        assert_eq!(msg_1, msg_2);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_session_message_segment_req() {
+        let frame_info = FrameInfo {
+            frame_id: 10,
+            total_segments: 255,
+            missing_segments: bitarr![1; 256],
+            last_update: SystemTime::now(),
+        };
+
+        let msg_1 = SessionMessage::<466>::Request(SegmentRequest::from_iter(vec![frame_info]));
+        let data = Vec::from(msg_1.clone());
+        let msg_2 = SessionMessage::try_from(&data[..]).unwrap();
+
+        assert_eq!(msg_1, msg_2);
+
+        match msg_1 {
+            SessionMessage::Request(r) => {
+                let missing_segments = r.into_iter().collect::<Vec<_>>();
+                let expected = (0..=7).map(|s| SegmentId(10, s)).collect::<Vec<_>>();
+                assert_eq!(expected, missing_segments);
+            }
+            _ => panic!("invalid type"),
+        }
+    }
+
+    #[test]
+    fn test_session_message_frame_ack() {
+        let mut rng = thread_rng();
+        let frame_ids: Vec<u32> = (0..500).map(|_| rng.gen()).collect();
+
+        let msg_1 = SessionMessage::<466>::Acknowledge(frame_ids.into());
+        let data = Vec::from(msg_1.clone());
+        let msg_2 = SessionMessage::try_from(&data[..]).unwrap();
+
+        assert_eq!(msg_1, msg_2);
+    }
+
+    #[test]
+    fn session_message_segment_request_bitset_test() {
+        let seg_req = SegmentRequest::<466>([(10, 0b00100100)].into());
+
+        let mut iter = seg_req.into_iter();
+        assert_eq!(iter.next(), Some(SegmentId(10, 2)));
+        assert_eq!(iter.next(), Some(SegmentId(10, 5)));
+        assert_eq!(iter.next(), None);
```

Comment:
```suggestion
        assert_eq!(vec![SegmentId(10, 2), SegmentId(10, 5)], seg_req.into_iter().collect::<Vec<SegmentId>>());
```

Reply:
I'd like to keep as is, so it is more explicit.